src/database-controller/src/poller/utils.js

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+require('module-alias/register');
+const _ = require('lodash');
+const { v4: uuidv4 } = require('uuid');
+
+function isUnrecoverableResponse(response) {
+  // indentify if the response is recoverable
+  const statusCode = _.get(response, 'statusCode', 0);
+  const message = _.get(response, 'body.message', '');
```

Comment:
If the error is connection reset or connection eof error, should return false

Reply:
Such as 
 // Treat Permanent Framework Error as Permanent Error only if it must be
 // Permanent Error.
 if apiErrors.IsInvalid(apiErr) ||
  apiErrors.IsRequestEntityTooLargeError(apiErr) {
  // TODO: Also check net.IsConnectionRefused
  if net.IsConnectionReset(apiErr) || net.IsProbableEOF(apiErr) {
   // The ApiServer Permanent Error may be caused by Network Transient Error,
   // so treat it as Unknown Error for safety.
   return PodMatchResult{
    CodeInfo:    completionCodeInfoMap[CompletionCodePodCreationUnknownError],
    Diagnostics: diag,
   }
  } else {
   return PodMatchResult{
    CodeInfo:    completionCodeInfoMap[CompletionCodePodCreationPermanentError],
    Diagnostics: diag,
   }
  }
 }

---
In reply to: [570742486](https://github.com/microsoft/pai/pull/5284#discussion_r570742486) [](ancestors = 570742486)

Reply:
I did some investigation on this issue. There is no need to modify the code here.

First, k8s node client uses `request` to handle the HTTP request:

https://github.com/kubernetes-client/javascript/blob/61f8d3640485a177f1785601ba05a9884b6c34d8/src/gen/api/customObjectsApi.ts#L183-L193

For `request`, any network error will be raised directly. Thus there will be no valid `response`.

I also tested it by the following code:

```
const request = require('request');
request('https://download.pytorch.org/whl/cu101/torch-1.7.0%2Bcu101-cp36-cp36m-linux_x86_64.whl', function (error, response, body) {
  console.error('error:', error); // Print the error if one occurred
  console.error('response:', response)
 
});
```

If we cut the network connection during the request, the output will be:

```
error: { Error: read ECONNRESET
    at TCP.onread (net.js:622:25) errno: 'ECONNRESET', code: 'ECONNRESET', syscall: 'read' }
response: undefined
```

Reply:
I mean apiserver may receive invalid json due to network issue, and return 422 to u

Reply:
This shouldn't happen in node.js because any network error will be raised directly.

Reply:
See https://github.com/microsoft/pai/issues/4840
The nodejs k8s client may behavior good, if it will not try decode body when connection err

Reply:
You can find the logic here: https://github.com/kubernetes-client/javascript/blob/61f8d3640485a177f1785601ba05a9884b6c34d8/src/gen/api/customObjectsApi.ts#L412-L421

If there's any error, the node.js k8s client will call `reject(error);`, which means the error will be raised.

It starts to decode the body only when there is no error detected.

Reply:
ok